File: shared/src/commonMain/kotlin/com/vnidrop/app/feature/saveddevices/SavedDeviceExperienceModels.kt
package com.vnidrop.app.feature.saveddevices

import com.vnidrop.app.core.PendingTargetedOfferModel
import com.vnidrop.app.core.TargetedTransferStateModel

sealed interface PairingPrompt {
	data class Eligibility(val peerEndpointId: String, val remoteDisplayName: String?) : PairingPrompt

	data class IncomingRequest(val peerEndpointId: String, val remoteDisplayName: String?) : PairingPrompt
}

data class PairingPromptState(
	val prompt: PairingPrompt? = null,
	val busy: Boolean = false,
)

data class TargetedOfferState(
	val pending: List<PendingTargetedOfferModel> = emptyList(),
	val senderDisplayNames: Map<String, String> = emptyMap(),
	val respondingIds: Set<String> = emptySet(),
) {
	val current: PendingTargetedOfferModel?
		get() = pending.firstOrNull()

	val currentSenderDisplayName: String?
		get() = current?.senderEndpointId?.let(senderDisplayNames::get)
}

enum class SavedDeviceTransferDirection {
	Outgoing,
	Incoming,
}

enum class SavedDeviceTransferAction {
	Receive,
	Resume,
	Cancel,
	Delete,
}

data class SavedDeviceTransferItem(
	val id: String,
	val peerEndpointId: String,
	val peerDisplayName: String?,
	val direction: SavedDeviceTransferDirection,
	val transferName: String,
	val fileCount: ULong,
	val totalSize: ULong,
	val verifiedBytes: ULong,
	val state: TargetedTransferStateModel,
	val createdAt: Long,
	val updatedAt: Long,
	val availableActions: List<SavedDeviceTransferAction>,
	val progressFraction: Float?,
)

internal fun savedDeviceTransferActions(
	direction: SavedDeviceTransferDirection,
	state: TargetedTransferStateModel,
): List<SavedDeviceTransferAction> = buildList {
	if (direction == SavedDeviceTransferDirection.Incoming) {
		when (state) {
			TargetedTransferStateModel.Approved -> add(SavedDeviceTransferAction.Receive)
			TargetedTransferStateModel.Interrupted -> add(SavedDeviceTransferAction.Resume)
			else -> Unit
		}
	}
	when (state) {
		TargetedTransferStateModel.Preparing,
		TargetedTransferStateModel.Offering,
		TargetedTransferStateModel.AwaitingApproval,
		TargetedTransferStateModel.Approved,
		TargetedTransferStateModel.Connecting,
		TargetedTransferStateModel.Transferring,
		TargetedTransferStateModel.Interrupted -> add(SavedDeviceTransferAction.Cancel)
		TargetedTransferStateModel.Completed,
		TargetedTransferStateModel.Declined,
		TargetedTransferStateModel.Cancelled,
		TargetedTransferStateModel.Failed -> add(SavedDeviceTransferAction.Delete)
		TargetedTransferStateModel.Deleted -> Unit
	}
}

internal fun savedDeviceTransferProgress(
	direction: SavedDeviceTransferDirection,
	state: TargetedTransferStateModel,
	verifiedBytes: ULong,
	totalSize: ULong,
): Float? {
	if (direction != SavedDeviceTransferDirection.Incoming || totalSize == 0UL) return null
	if (state !in progressStates) return null
	return (verifiedBytes.toDouble() / totalSize.toDouble()).coerceIn(0.0, 1.0).toFloat()
}

private val progressStates = setOf(
	TargetedTransferStateModel.Connecting,
	TargetedTransferStateModel.Transferring,
	TargetedTransferStateModel.Interrupted,
)

File: apple/VniDrop/Features/SavedDevices/SavedDeviceExperienceModels.swift
import Foundation

/// Presentation-level saved-device models, ported from
/// `feature/saveddevices/SavedDeviceExperienceModels.kt`.

/// The one consent question currently worth asking. Only one is shown at a time:
/// an incoming request always outranks an eligibility we could act on ourselves.
enum PairingPrompt: Equatable, Identifiable {
	/// We completed a qualifying transfer and may ask this peer to pair.
	case eligibility(peerEndpointId: String, remoteDisplayName: String?)
	/// This peer asked us; we approve or decline.
	case incomingRequest(peerEndpointId: String, remoteDisplayName: String?)

	var peerEndpointId: String {
		switch self {
		case .eligibility(let id, _), .incomingRequest(let id, _): return id
		}
	}

	var remoteDisplayName: String? {
		switch self {
		case .eligibility(_, let name), .incomingRequest(_, let name): return name
		}
	}

	var id: String {
		switch self {
		case .eligibility(let id, _): return "eligibility-\(id)"
		case .incomingRequest(let id, _): return "incoming-\(id)"
		}
	}
}

struct PairingPromptState: Equatable {
	var prompt: PairingPrompt?
	var busy = false
}

struct TargetedOfferState: Equatable {
	var pending: [PendingTargetedOfferModel] = []
	/// Display names for senders we already have saved. A sender we have not
	/// saved has no trustworthy name, so the UI falls back to a generic label
	/// rather than rendering a peer-supplied string as if it were verified.
	var senderDisplayNames: [String: String] = [:]
	var respondingIds: Set<String> = []

	var current: PendingTargetedOfferModel? { pending.first }

	var currentSenderDisplayName: String? {
		guard let current else { return nil }
		return senderDisplayNames[current.senderEndpointId]
	}
}

enum SavedDeviceTransferDirection: Equatable, Sendable {
	case outgoing
	case incoming
}

enum SavedDeviceTransferAction: Equatable, Sendable {
	case receive
	case resume
	case cancel
	case delete
}

/// One targeted transfer as the details surface renders it: resolved against the
/// local endpoint so "peer" means the *other* device, whichever side we are on.
struct SavedDeviceTransferItem: Equatable, Identifiable, Sendable {
	let id: String
	let peerEndpointId: String
	let peerDisplayName: String?
	let direction: SavedDeviceTransferDirection
	let transferName: String
	let fileCount: UInt64
	let totalSize: UInt64
	let verifiedBytes: UInt64
	let state: TargetedTransferStateModel
	let createdAt: Int64
	let updatedAt: Int64
	let availableActions: [SavedDeviceTransferAction]
	let progressFraction: Double?

	init(
		id: String,
		peerEndpointId: String,
		peerDisplayName: String?,
		direction: SavedDeviceTransferDirection,
		transferName: String,
		fileCount: UInt64,
		totalSize: UInt64,
		verifiedBytes: UInt64,
		state: TargetedTransferStateModel,
		createdAt: Int64,
		updatedAt: Int64
	) {
		self.id = id
		self.peerEndpointId = peerEndpointId
		self.peerDisplayName = peerDisplayName
		self.direction = direction
		self.transferName = transferName
		self.fileCount = fileCount
		self.totalSize = totalSize
		self.verifiedBytes = verifiedBytes
		self.state = state
		self.createdAt = createdAt
		self.updatedAt = updatedAt
		self.availableActions = Self.actions(direction: direction, state: state)
		self.progressFraction = Self.progress(
			direction: direction,
			state: state,
			verifiedBytes: verifiedBytes,
			totalSize: totalSize
		)
	}

	private static func actions(
		direction: SavedDeviceTransferDirection,
		state: TargetedTransferStateModel
	) -> [SavedDeviceTransferAction] {
		var actions: [SavedDeviceTransferAction] = []
		if direction == .incoming {
			switch state {
			case .approved: actions.append(.receive)
			case .interrupted: actions.append(.resume)
			default: break
			}
		}
		switch state {
		case .preparing, .offering, .awaitingApproval, .approved,
				.connecting, .transferring, .interrupted:
			actions.append(.cancel)
		case .completed, .declined, .cancelled, .failed:
			actions.append(.delete)
		case .deleted:
			break
		}
		return actions
	}

	private static func progress(
		direction: SavedDeviceTransferDirection,
		state: TargetedTransferStateModel,
		verifiedBytes: UInt64,
		totalSize: UInt64
	) -> Double? {
		guard direction == .incoming, totalSize > 0 else { return nil }
		guard state == .connecting || state == .transferring || state == .interrupted else {
			return nil
		}
		return min(1, Double(verifiedBytes) / Double(totalSize))
	}
}
